File: Leetcode/Leetcode-DSA-course/Trees-and-Graphs/1092-maximum-difference-between-node-and-ancestor/maximum-difference-between-node-and-ancestor.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 * int val;
 * TreeNode left;
 * TreeNode right;
 * TreeNode() {}
 * TreeNode(int val) { this.val = val; }
 * TreeNode(int val, TreeNode left, TreeNode right) {
 * this.val = val;
 * this.left = left;
 * this.right = right;
 * }
 * }
 */
class Solution {
    int ans = 0;

    public int maxAncestorDiff(TreeNode root) {
        dfs(root, root.val, root.val);
        return ans;
    }

    void dfs(TreeNode node, int currMax, int currMin) {
        if (node == null)
            return;
        currMax = Math.max(currMax, node.val);
        currMin = Math.min(currMin, node.val);
        if (node.left == null && node.right == null) {
            ans = Math.max(currMax - currMin, ans);
            return;
        }
        dfs(node.left, currMax, currMin);
        dfs(node.right, currMax, currMin);
    }
}

File: Leetcode/Leetcode-DSA-course/Trees-and-Graphs/1092-maximum-difference-between-node-and-ancestor/maximum-difference-between-node-and-ancestor.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def maxAncestorDiff(self, root: Optional[TreeNode]) -> int:
        ans = 0

        def dfs(node, currMax, currMin):
            nonlocal ans
            if not node:
                return
            currMax = max(currMax, node.val)
            currMin = min(currMin, node.val)
            if node.left == None and node.right == None:
                ans = max(ans, currMax - currMin)
                return
            dfs(node.left, currMax, currMin)
            dfs(node.right, currMax, currMin)

        dfs(root, root.val, root.val)
        return ans
